### ISSUE TITLE
For discussion: add a minimum viable service to bluesky RunEngine itself

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky#1555]
The [bluesky-queueserver](https://github.com/bluesky/bluesky-queueserver) and associated projects have staked out a good scope for a full-featured "RunEngine as a service". To use it, users have to take a discontinuous jump from the traditional REPL-based (typically IPython-based) control to remote control from a client. There is also some configuration of services required.

There may be room for an intermediate scope, where users can continue to operate as before, directly on the `RunEngine` instance in a REPL, while also sending requests from a client. I sketched out an [ugly version of this](https://github.com/danielballan/RE-0MQ), with some help from @tacaswell (who deserves no blame for its ugliness) just to get us thinking.

A better version of that sketch might look like this:

```py
RE = RunEngine(...)  # no change here
RE.listen("127.0.0.1:7070", api_key=os.environ["RE_API_KEY"])
```

That last command would start a dead-simple `aiohttp` server running on the same captive event loop as the RunEngine. It would accept commands over HTTP like:

```
curl -X POST -H 'Authorization:Apikey SECRET' -d '{...}' https://127.0.0.1:7070/...
```

The server would enforce simple API key authentication, but there would be no user model or notion of authorization or scopes.

(Alternatively, it could be a 0MQ server. We should probably pick one, I think, but either technology could serve the purpose.) The payload `{...}` would be a JSON blob similar to (likely a subset of) the JSON API used by queueserver. (Perhaps the protocol code could be moved from wherever it lives in the queueserver repos into `bluesky` itself.) Now, this would _not_ provide a queue or any persistence. It would only accept one plan at a time; it would reject requests if busy.

This is, of course, similar to moving the "RE worker" of queueserver into bluesky core and shipping it as a default. It feels like a natural evolution, now that we understand the requirements of "RE as a service" better, with experience. I like that this proposal enables:

* An easy path to running a service for _trusted_ clients to access the RunEngine with no additional configuration or processes
* Concurrent usage in the REPL-based mode and the remote/queue-based mode, with dead simple conflict resolution (reject competing requests)

**Comment by whs92** [comment]
As @dmgav stated on slac, once you provide all the nescesary features like submission and checking of plans and devices, monitoring of running plans, etc you'll  probably end up with the qsever. 

Given that, would an approach be to use the existing qserver, but make an IPython REPL client for it that:

- Connected to the qserver (something like `RE.connect(host:port)`)
- Allowed for some control of the RE worker environment (something like `RE.start_re_worker()`)
- Recieved the list of allowed plans and devices and created some way of allowing tab auto-completion/ inspection of these in IPython environment
- Allowed for an almost identical interface to the existing submission of plans like `RE(scan([det],motor, 1,2,10))`.
- [immediately ran submitted plans](https://blueskyproject.io/bluesky-queueserver/using_queue_server.html#immediate-execution-of-plans) in the queue, or waited if the queue was busy
- and then subscribe to the document stream from the RE worker to plot and stuff as before. The client could wait if there were other plans running in the queue.

I think I agree with what @danielballan said on slac that most users will eventually need a queue, I am just not sure that they will all want to know about it or interact with it. Creating and modifying a queue makes sense in a GUI, but I think users of the IPython interface are used to and like the REPL style. Reordering and inspection of a queue in IPython feels clunky. If you did make an IPython client for the queue-server, then like I say above you'd still have to interact with a queue, but the default could be immediate processing or otherwise just wait for the queue to finish if some other client has submitted a plan. 

For me the benefit of creating this other client to the qserver is to allow users who want it a REPL based interface, but without me having to maintain two environments. The queue-server is great because it allows people outside the experiment to see what's currently running, what's been run in the past, and what's going to be run in the future. It also allows me to control who can do what, this is particuarly attractive for me because I have multiple users and user profiles for one experimental resource.

**Comment by danielballan** [comment]
That is a useful perspective and something I think I have also heard our scientists ask for: a way to attach to the queue with something that feels like RE to use. There are some technical challenges there: Does it accept plans or plan names? What about devices? But I think we can and should work that out.

This is coming from the perspective of “I have a working environment and I am not ready to radically change it, but I would like to do some GUI development in parallel and be able to submit plans through it.” Even if that is only a transient state on the path to full queue sever adoption, a state that users basically always need to advance through, it may help to make the path smoother.

**Comment by danielballan** [comment]
Our current answer for (1) is of course: https://github.com/bluesky/bluesky-queueserver-api

**Comment by dmgav** [comment]
The idea of implementing RE-compatible interface existed from the very beginning, but no viable solutions were proposed, mostly because the objects, such as devices and plans, exist in the (remote) worker environment and are not directly accessible to the client. So calls like `RE(count([det], num=10)` can not be implemented on the client side (unless we create 'dummy' objects for available plans and devices in 'user_ns', which may be confusing for general user).

**Comment by callumforrester** [comment]
This looks like an interesting idea! Will definitely decrease the likelihood of divergence between different facilities. I have a few thoughts but I'm not 100% about them so welcome discussion. 

I have reservations about a request/response API (i.e. REST or similar) for Bluesky. The typical usage model of a RunEngine in my experience is more asynchronous; a few requests interspersed with regular updates e.g.
![bs-usage](https://user-images.githubusercontent.com/29771545/207273718-98726dbd-22fd-42cf-b53d-83214d487fca.png)

Getting these updates via polling seems wasteful and not the right fit given that most things using the RE will be fundamentally event driven, which is why I was looking at using a message broker instead. 0MQ might be a better fit...? I haven't had much experience with it. Websockets are also a thought.

Perhaps the best way to encourage the various Bluesky-using facilities to make use of this is to ship it sans-IO and provide easy ways for people to bolt on their own. Plus maybe a few defaults with optional dependencies. So you would get something like:

```python
# Import just has Bluesky core dependencies
from bluesky import RunEngine
# Import requires optional dependencies: fastapi and uvicorn
from bluesky.server.rest import RestListener
# Import requires optional dependency: stomp.py
from bluesky.server.stomp import StompListener

RE = RunEngine()
RE.listen(RestListener("0.0.0.0", 80))
# ...or
RE.listen(StompListener("messagebroker.myfacility.org", 61616))
```

**Comment by coretl** [comment]
> @callumforrester: to ship it sans-IO 

I like the idea of sans-IO, will make it more likely to fit with whatever people want to use in production

> @whs92: Recieved the list of allowed plans and devices and created some way of allowing tab auto-completion/ inspection of these in IPython environment

I agree with this, and with @untzag's calls for self-describing plans. What I would like out of such an interface is a way to:
- Ask for the available plans
- Ask for the argument names, types and descriptions for a given plan
- Ask for the objects of a given type that exist in the RE
- Run a plan with arguments as per @callumforrester's diagram

**Comment by danielballan** [comment]
Thanks @callumforrester. Since Diamond’s use case is distinct, I think if we can engage with it in this design process we will significantly improve our chances of choosing a strong abstraction.

I like the sans-IO idea very much. And I think you are right that HTTP is probably not a good fit for this communication channel, both for the reason you mention and others. (To be clear, exposing it coarsely _behind_ HTTP, as we do in queuesever, is a separate question.)

Let me describe our use case more. We have nearly 30 instruments now running with Bluesky in IPython. Some of the Bluesky users (by which I mean instrument staff) have prototyped their own Qt-based GUIs that run in-process, launched from IPython using its integrated Qt event loop. We dislike this pattern because it too tightly couples the GUI to the (EPICS) control process. It is fragile and hard to test.

But an upgrade to queuesever requires accepting a discontinuous change in their process and a certain amount of risk. First, because their IPython configurations often have grown some assumptions that queuesever breaks, some expert help is usually required to make the configuration compatible with queuesever. This is technically straightforward, but any modifications to those configurations, especially ones that the users do not completely understand, make users wary. Second, the RE is moved into a headless process into which users have less direct visibility and control.

if users wish to have both modes operation available during the transition, there is currently no fluid way to provide this. Hence the interest in an intermediate step that enables the old way and the new way coexist more seamlessly, at least for some time.

Finally, a question for this group: At this point I do not think I understand whether we need self-description for all use cases. A lot of the GUIs we see on the floor assume _a priori_ knowledge of a handful of plans and devices. If important plans or devices are added or removed, the GUI is updated. All we would really require to move these GUIs out of process is the basic protocol that @callumforrester outlined above. Adding a protocol to describe the environment seems like a separable step that may add significant complexity. More importantly, if changes are needed to the IPython configuration to support it, it will add adoption overhead and risk. (I want to emphasize: even if it is not required for our use case, we could include it as an optional feature to support others’ use cases.) But I would like to understand better what this is most needed for. Client-side validation? Automatic GUI generation? Something else?

**Comment by coretl** [comment]
> But I would like to understand better what this is most needed for. Client-side validation? Automatic GUI generation? Something else?

Mainly for automatic GUI generation, which may also do some client-side validation. We have 2 broad categories of plans:
1. The day-to-day plans the beamlines always run. These very rarely change, so a custom GUI is reasonable to write, so they can move this button just HERE, and prompt if the user did something silly.
2. The experimental plans that they have JUST written for the user. They can prototype in a tight loop within IPython, but then they want the user to queue up a bunch of scans to run overnight. A simple form-based autogenerated UI that lets them introspect the parameters and hit the Queue button would let them do that. The plan may be thrown away the next day, so a custom GUI is overkill.

1 should work without any introspection, so I think these type hints should be optional. There is then the incentive to add them to make 2 work, so the autogenerated GUI can be used for plans.

It's good to note here that these are all aspirations, we have no idea what our plans will actually end up looking like, especially when we start considering flyscans, so I think a route for adding 2 later, but considering how it might fit in would be best.

**Comment by callumforrester** [comment]
I think I would be interested in all the use cases you list, @danielballan. But, as you say, they add complexity so perhaps they belong in another module. That's potentially fine if the hooks are available to add such components in an easy way. 

I think a very important use case for environment introspection is debugging. I feel like if I'm setting my RunEngine up in a headless process on a new beamline, or have just written a new plan to run for the first time, I would like a set of simple commands I can run to sanity check that everything is as I think it is (e.g. can the RunEngine actually see my new plan?). 

I think writing the code to enable this would be very worthwhile, whether or not it is shipped by default, because rather than adding complexity we are moving it from commissioning time to development time, which is perhaps preferable because you have fewer users shouting at you during development time.

**Comment by danielballan** [comment]
Thanks, that is all very helpful, and it makes sense to me.

Even for our simple use cases, I can imagine wanting to sanity check, "Do these names exist in the RE namespace (and do they have the expected types)?" so that we can show a clear error dialog early if someone commented-out or renamed a motor we were expecting to find.

As an MVP, would it be suitable to have a response that provides a dump of public variables in `globals()`, their types, and (if callable) their signatures? If combined with the long-planned #1476 this might be sufficient for many use cases, and it would require zero configuration.

**Comment by danielballan** [comment]
An unrelated observation: the [Jupyter protocol](https://jupyter-protocol.readthedocs.io/en/latest/messaging.html) uses five 0MQ channels, each for a different function (heartbeat, control, shell, stdin, IOpub). I believe queuesever now has two: the document steam and the control channel. (And maybe another for logs?)

I like the idea of multiple sans-IO channels with a high-level convenience layer that lets me type one short line to get a basic setup with 0MQ. I think that bluesky’s long-standing [solution for 0MQ document publishing](https://blueskyproject.io/bluesky/callbacks.html#subscriptions-in-separate-processes-or-host-with-0mq) could be incorporated into one streamlined step, similarly to how dask does it.

**Comment by untzag** [comment]
> At this point I do not think I understand whether we need self-description for all use cases.

I 100% agree that plan annotations are not always needed and should never be required for running a plan. Speaking personally, I just want to see a standardized way to serialize plan signatures so that tools around the peripheral can interoperate based on that standard.

Analogy to Happi support. Over-engineered for static systems with highly customized interfaces. Great for dynamic systems and tooling.

**Comment by untzag** [comment]
> if users wish to have both modes operation available during the transition, there is currently no fluid way to provide this. Hence the interest in an intermediate step that enables the old way and the new way coexist more seamlessly, at least for some time.

I would consider building this as a temporary shim only instead of something for general consumption. I can imagine writing a `queueserver-api` wrapper that quacks like a RE locally, but such an interface would need to rely on a bunch of heuristics. As @dmgav points out, it also would need to rely on assumptions about identical devices server side and locally. You might get lucky with such assumptions for your 30 existing instruments, but they will fall down with the silly stuff I'm doing.

Regarding this idea of coexistence, I suggest that you really want to build a shim that plugs into the full queueserver. The new clients you'll want to start prototyping will need all the rest of the queueserver API. You would just need to figure out how this shim will work with other scans. You could wait, or interrupt. All your varied Qt apps could then add queueserver support gradually to work towards your ultimate goal.

Just my thoughts, often wrong!

**Comment by danielballan** [comment]
I don’t think a queueserver-api shim would address the costs risks and costs of making that discontinuous jump from the current in-process operation to the queueserver-based operation. The primary goal is to provide a stepping stone to remote operation that enables coexistence, an add-on to the existing deployment and usage with zero configuration or profile modification.

**Comment by danielballan** [comment]
Notes from an in-person chat with @tacaswell:

The ability of the RunEngine to _become_ a server is a good success story of something developed out of core (in queue server) moving into core and becoming a little more public. What we've learned from Diamond is there is more than one way to have a queue, and it might not even be in Python, so it is good to have a public interface that other queues can use.

Let's start by having an explicit hand-off of control that blocks:

```
RE.listen(...)
RE(scan(...))  # interactive local control
RE.remote_control()  # blocks for remote control until remote client releases it
RE(scan(...))  # back to local control
```

This avoids potentially confusing behavior, and sticky questions like what happens with `LiveTable` output in IPython during remote control.

**Comment by danielballan** [comment]
Summary of discussion with @dmgav:

The complexity of queueserver mostly relates to supporting fully remote or automated control with no one physically accessing the beamline. To support that, Dmitri has found these requirements:

* restart the worker environment
* customize the worker environment (i.e. upload custom plans or functions)
* sometimes execute code that is not a plan in the main thread and in background threads
* optionally, enforce granular access controls

The proposed addition to bluesky core does not do any of that. Its goal is not fully remote control. Rather, it aims to support dual control from IPython and from a separate process (GUI or Terminal app). Currently, beamline staff tend to write GUIs that run directly inside IPython, and we want to enable them (and ourselves) to write GUIs that connect to the RunEngine from out of process. The connection will be encrypted (using the built-in 0MQ mechanism) but will not be intended to serve users outside the local network or to provided granular access control. Those features are still the domain of queueserver.

The _sans I/O_ JSON API used by the proposed addition to bluesky core should be a simple as possible. We should coordinate it with the JSON API used by queueserver, using the same names where practical, so that given GUI can easily be modified to communicate with one or the other.

@dmgav is game to write a proposed JSON API, a simplified version of the queueserver worker API, and then collect comments. That will be the next step.
